reframe/core/variables.py

```diff
@@ -0,0 +1,217 @@
+    classes of the target class. After this, the VarSpace is extended with
+    the information from the local variable space, which is stored under the
+    target class' attribute ``_rfm_local_var_space``. If no target class is
+    provided, the VarSpace is simply initialized as empty.
+    '''
+
+    localAttrSpaceName = '_rfm_local_var_space'
+    localAttrSpaceCls = LocalVarSpace
+    attrSpaceName = '_rfm_var_space'
+
+    def join(self, other):
+        '''Join an existing VarSpace into the current one.'''
+        for key, var in other.items():
+
+            # Make doubly declared vars illegal. Note that this will be
+            # triggered when inheriting from multiple RegressionTest classes.
+            if key in self.vars:
+                raise ValueError(
+                    f'attribute {key!r} is declared in more than one of the '
+                    f'parent classes'
+                )
+
+            self.vars[key] = var
+
+    def extend(self, cls):
+        '''Extend the VarSpace with the content in the LocalVarSpace.
+
+        Merge the VarSpace inherited from the base classes with the
+        LocalVarSpace. Note that the LocalVarSpace can also contain
+        define and undefine actions on existing vars. Thus, since it
+        does not make sense to define and undefine a var in the same
+        class, the order on which the define and undefine functions
+        are called is not preserved. In fact, applying more than one
+        of these actions on the same var for the same local var space
+        is disallowed.
+        '''
+        localVarSpace = getattr(cls, self.localAttrSpaceName)
+
+        # Extend the VarSpace
+        for key, var in localVarSpace.items():
+
+            # Disable redeclaring a variable
+            if key in self.vars:
+                raise ValueError(
+                    f'cannot redeclare a variable ({key})'
+                )
+
+            self.vars[key] = var
+
+        # Undefine the vars as indicated by the local var space
+        for key in localVarSpace.undefined:
+            self._check_var_is_declared(key)
+            self.vars[key].undefine()
+
+        # Define the vars as indicated by the local var space
+        for key, val in localVarSpace.definitions.items():
+            self._check_var_is_declared(key)
+            self.vars[key].define(val)
+
+    def _check_var_is_declared(self, key):
+        if key not in self.vars:
+            raise ValueError(
+                f'var {key!r} has not been declared'
+            )
+
+    def insert(self, obj, cls):
+        '''Insert the vars in the regression test.
+
+        :param obj: The test object.
+        :param cls: The test class.
+        '''
+
+        for name, var in self.items():
+            setattr(cls, name, var.field(*var.types))
+            getattr(cls, name).__set_name__(obj, name)
+
+            # If the var is defined, set its value
+            if not var.is_undef():
```

Comment:
I probably prefer this.

```suggestion
            if var.is_defined():
```